src/libexpr/paths.cc

```diff
@@ -0,0 +1,93 @@
+#include "eval.hh"
+#include "util.hh"
+#include "fs-input-accessor.hh"
+
+namespace nix {
+
+SourcePath EvalState::rootPath(const Path & path)
+{
+    return {rootFS, CanonPath(path)};
+}
+
+void EvalState::registerAccessor(ref<InputAccessor> accessor)
+{
+    inputAccessors.emplace(accessor->number, accessor);
+}
+
+std::string EvalState::encodePath(const SourcePath & path)
+{
+    /* For backward compatibility, return paths in the root FS
+       normally. Encoding any other path is not very reproducible (due
+       to /nix/store/virtual000...<N>) and we should deprecate it
+       eventually. So print a warning about use of an encoded path in
+       decodePath(). */
+    return path.accessor == rootFS
+        ? path.path.abs()
+        : fmt("%s%08x-source%s", virtualPathMarker, path.accessor->number, path.path.absOrEmpty());
+}
+
+SourcePath EvalState::decodePath(std::string_view s, PosIdx pos)
+{
+    if (!hasPrefix(s, "/"))
+        throwEvalError(pos, "string '%1%' doesn't represent an absolute path", s);
+
+    if (hasPrefix(s, virtualPathMarker)) {
+        auto fail = [s]() {
+            throw Error("cannot decode virtual path '%s'", s);
+        };
+
+        s = s.substr(virtualPathMarker.size());
+
+        try {
+            auto slash = s.find('/');
+            if (slash == std::string::npos) fail();
+            size_t number = std::stoi(std::string(s, 0, slash), nullptr, 16);
+            s = s.substr(slash);
+
+            auto accessor = inputAccessors.find(number);
+            if (accessor == inputAccessors.end()) fail();
+
+            SourcePath path {accessor->second, CanonPath(s)};
+
+            warn("applying 'toString' to path '%s' and then accessing it is deprecated, at %s", path, positions[pos]);
```

Comment:
Why should this be deprecated? This PR currently breaks a function like this (which yes, without context doesn't make much sense):

```
nix-repl> f = path: builtins.readDir (toString path)

nix-repl> normalPath = ./doc    

nix-repl> f normalPath          
{ manual = "directory"; }

nix-repl> lazyPath = (builtins.fetchTree { type = "git"; url = ./.; }).outPath 

nix-repl> f lazyPath                                                           
error: cannot decode virtual path '/nix/store/virtual0000000000000000000000007-source'

nix-repl> f (lazyPath + "/doc")
warning: applying 'toString' to path '/home/tweagysil/src/nix/doc' and then accessing it is deprecated, at «string»:1:8
{ manual = "directory"; }
```

It's very odd how it's broken when accessing the direct path, but works (though deprecated when a subpath), but I think this can be fixed.

Since we can return the correct path though, why should this functionality be deprecated? Are there any problems with allowing this?

Reply:
I do this all the time.

Nix has a habit of recursively copying `path` types and its way too easy to accidentally copy a 2Gb tree because someone wrote `builtins.trace "reading files in: ${dir}"`.

Because of that basically every interface and lib routine I write converts path to string, and keeps it a string until eventually using it.

Strings are "less efficient" sure, but spuriously copying paths to the Nix store is worse, so it's a trade I stand by.

I'm going to have a lot of refactoring to do if `toString ./.` is deprecated...

Reply:
@infinisil I've fixed the "cannot decode virtual path" error.